### ISSUE TITLE
Extra parameters for FunctionExecutor to pass to execute()

### ISSUE THREAD

**Issue by himekifee** [sashirestela/simple-openai#200]
First of all thanks for your hard work on the projects. It really helped me build my Minecraft Mod with AI equipped. I just added functionality to ask AI to generate a function calling to execute server commands, which requires some context about the sender and the server itself, essentially requiring passing external parameters to the `execute()` function. One alternative way is to add some information to the prompt and hope the AI fill in proper arguments which is of cause only capable of processing text. I wonder if there is a way to pass objects to `execute()`? Thanks.

**Comment by sualeh** [comment]
I think that this is an important requirement. The \https://github.com/TheoKanning/openai-java allowed it by actually storing an instance of the function executor class (see https://github.com/TheoKanning/openai-java/blob/main/api/src/main/java/com/theokanning/openai/completion/chat/ChatFunction.java#L33) which could have its own internal state. In https://github.com/sashirestela/simple-openai on the other hand, all that is stored is the class (instead of an instance of the class) that can execute a function.

**Comment by sualeh** [comment]
Contrast:
https://github.com/TheoKanning/openai-java/blob/main/api/src/main/java/com/theokanning/openai/completion/chat/ChatFunction.java#L55-L59
with 
https://github.com/sashirestela/simple-openai/blob/8c46ce7bef6c946a18f6a94fb10fee16792e4da5/src/main/java/io/github/sashirestela/openai/common/function/FunctionDef.java#L18

**Comment by sualeh** [comment]
Another way to handle this could be simply return an instantiated Functional object without this line of code, and leave the execution up to the calling code. This may be a much simpler short-term solution.

https://github.com/sashirestela/simple-openai/blob/8c46ce7bef6c946a18f6a94fb10fee16792e4da5/src/main/java/io/github/sashirestela/openai/common/function/FunctionExecutor.java#L84

**Comment by sualeh** [comment]
Until then, I am going to have to instantiate my own Functional and execute it in my own code, so I can provide the required context.

**Comment by dyeray** [comment]
I think we are on the same case. I think the whole function implementation can be rethought. We wanted to pass a parameter to the functional provider like you, but also, we wanted the provider to be a suspend function, since we are using Kotlin coroutines, plus we wanted to be able to use injection, plus do a proper separation of data/code (for now the library forces to use the same class for both).

So basically we implemented our own FunctionDef and FunctionExecutor. Luckily the tool allows to easily replace it, as the client library only depends on Tool
